lib/neovim-plugin.nix

```diff
@@ -94,25 +99,90 @@ with lib;
+            inherit originalName;
+            lazyLoadDefaults =
+              (optionalAttrs (isColorscheme && colorscheme != null) { inherit colorscheme; }) // lazyLoadDefaults;
+          };
+        }
         // extraOptions;
 
       config =
         let
-          cfg = config.${namespace}.${name};
           extraConfigNamespace = if isColorscheme then "extraConfigLuaPre" else "extraConfigLua";
+          lazyLoaded = cfg.lazyLoad.enable or false;
+
+          # lz-n lazyLoad backend
+          # Transform plugin into attrset and set optional to true
+          # See `tests/test-sources/plugins/pluginmanagers/lz-n.nix`
+          optionalPlugin =
+            x:
+            if isColorscheme then x else ((if x ? plugin then x else { plugin = x; }) // { optional = true; });
+          mkFn = str: if str != "" then "function()\n" + str + "end" else null;
         in
         mkIf cfg.enable (mkMerge [
-          {
-            extraPlugins = (optional installPackage cfg.package) ++ extraPlugins;
-            inherit extraPackages;
-          }
-          (optionalAttrs callSetup {
-            ${extraConfigNamespace} = ''
-              require('${luaName}')${setup}(${optionalString (cfg ? settings) (helpers.toLuaObject cfg.settings)})
-            '';
-          })
+          # Always set
           (optionalAttrs (isColorscheme && (colorscheme != null)) { colorscheme = mkDefault colorscheme; })
-          (extraConfig cfg)
+
+          # Normal loading
+          (mkIf (!lazyLoaded) (mkMerge [
+            (extraConfig cfg)
+            {
+              extraPlugins = (optional installPackage cfg.package) ++ extraPlugins;
+              inherit extraPackages;
+            }
+            (optionalAttrs callSetup {
+              ${extraConfigNamespace} = ''
+                require('${luaName}')${setup}(${optionalString (cfg ? settings) (helpers.toLuaObject cfg.settings)})
+              '';
+            })
+          ]))
+
+          # Lazy loading with lz-n
+          (mkIf (lazyLoaded && config.plugins.lz-n.enable) (mkMerge [
+            (lib.removeAttrs (extraConfig cfg) [
+              "extraConfigLua"
+              "extraConfigLuaPre"
+            ])
+            {
+              extraPlugins = (optional installPackage (optionalPlugin cfg.package)) ++ extraPlugins;
+              inherit extraPackages;
+            }
+            {
+              plugins.lz-n.plugins = [
+                {
+                  __unkeyed-1 = packageName;
+                  after =
+                    if cfg.lazyLoad.after == null then
+                      mkFn (
+                        (extraConfig cfg).extraConfigLuaPre or ""
+                        + optionalString callSetup ''require('${luaName}')${setup}(${
+                          optionalString (cfg ? settings) (helpers.toLuaObject cfg.settings)
+                        }) ''
+                        + (extraConfig cfg).extraConfigLua or ""
+                      )
```

Comment:
Could this be the `after` option's `default`, instead of `null`? This `if then else` is _effectively_ setting a default.

Don't mind too much if it is set with `default` in mkOption, or if we use `lib.mkDefault` in a config definition. We should probably also have a `defaultText = literalExpression ""` too, for transparency.

Reply:
I guess this is more complex because `mkLazyLoadOption` isn't given an actual option path, so it can't (truly) know which `extraConfigLua` definitions are being used, etc. Same for `luaName`, `setup`, etc.

We can still _implement_ it as a `${namespace}.${name}.lazyload.after = mkDefault (mkFn "")`, we just would struggle to _document_ this in `defaultText`